### PR TITLE
[Wallet] Address book encapsulation.

### PR DESCRIPTION
**Pull request by furszy**
Essentially, have restricted `mapAddressbook` access to the wallet class only. Migrating every direct access to the correspondent getter, setter and iterator wrapper. Plus, solved a todo in `AddressTableModel::getAddressToShow` (df262e4).
 It's a preparation for the shielded addresses addressbook support.

### CODE REVIEW

**Review by random-zebra — APPROVED** [review]
Pretty nice. utACK f97ed1cd084e883f3be803c9b9f1f84487b1623f

**Comment by random-zebra** [comment]
Needs rebase

**Comment by furszy** [comment]
rebased

**Review by random-zebra — APPROVED** [review]
re-utACK 264d33106848dc176d24821e4f0d10f7aa266f1e

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -609,12 +615,15 @@ UniValue ListaddressesForPurpose(const std::string strPurpose)
     UniValue ret(UniValue::VARR);
     {
         LOCK(pwalletMain->cs_wallet);
-        for (const auto& addr : pwalletMain->mapAddressBook) {
-            if (addr.second.purpose != strPurpose) continue;
+        CAddressBookIterator it = pwalletMain->NewAddressBookIterator();
+        while (it.HasNext()) {
+            auto addrBook = it.GetValue();
+            if (addrBook.purpose != strPurpose) continue;
```
Comment (random-zebra) [inline src/wallet/rpcwallet.cpp:621]: Needs to be very careful with `continue` when changing a `for` loop to a `while`.
If the purpose doesn't match, we never reach `it.Next()`, therefore we have an infinite loop.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -2047,25 +2061,21 @@ UniValue ListReceived(const UniValue& params, bool by_label)
...
 
-    for (auto item_it = start; item_it != end; ++item_it)
-    {
-        const CTxDestination& address = item_it->first;
-        const std::string& label = item_it->second.name;
+    // Reply
+    UniValue ret(UniValue::VARR);
+    std::map<std::string, tallyitem> label_tally;
+
+    while(itAddrBook.HasNext()) {
```
Comment (random-zebra) [inline src/wallet/rpcwallet.cpp:2076]: Same here. There is a `continue` below (line 2081) that make this an infinite loop.
This is actually hit during `wallet_listreceivedby.py` (you can see travis consistently timing out).

**Review by random-zebra — CHANGES_REQUESTED** [review]
Found a bug with the addressbook iterator-based while loops.

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (furszy) [reply src/wallet/rpcwallet.cpp:621]: upps yeah, great catch.

**Comment by furszy** [comment]
updated following the feedback 👌

File: src/qt/addresstablemodel.cpp [hunk 3]
```diff
@@ -539,14 +539,9 @@ QString AddressTableModel::labelForAddress(const QString& address) const
...
-            LOCK(wallet->cs_wallet);
-            CTxDestination destination = DecodeDestination(address.toStdString());
-            std::map<CTxDestination, AddressBook::CAddressBookData>::iterator mi = wallet->mapAddressBook.find(destination);
-            if (mi != wallet->mapAddressBook.end()) {
-                return QString::fromStdString(mi->second.name);
-            }
-        }
+        LOCK(wallet->cs_wallet);
+        CTxDestination dest = DecodeDestination(address.toStdString());
+        return QString::fromStdString(wallet->GetNameForAddressBookEntry(dest));
```
Comment (random-zebra) [inline src/qt/addresstablemodel.cpp:544]: Why the LOCK? `GetNameForAddressBookEntry` already locks.

File: src/qt/walletmodel.cpp [hunk 4]
```diff
@@ -825,10 +824,7 @@ std::string WalletModel::getLabelForAddress(const CTxDestination& address)
     std::string label = "";
     {
         LOCK(wallet->cs_wallet);
-        std::map<CTxDestination, AddressBook::CAddressBookData>::iterator mi = wallet->mapAddressBook.find(address);
-        if (mi != wallet->mapAddressBook.end()) {
-            label = mi->second.name;
-        }
+        label = wallet->GetNameForAddressBookEntry(address);
```
Comment (random-zebra) [inline src/qt/walletmodel.cpp:827]: same here. No need for the LOCK block now.

File: src/qt/walletmodel.cpp [hunk 5]
```diff
@@ -291,12 +291,11 @@ bool WalletModel::updateAddressBookLabels(const CTxDestination& dest, const std:
...
-    std::map<CTxDestination, AddressBook::CAddressBookData>::iterator mi = wallet->mapAddressBook.find(dest);
-
+    auto optAdd = pwalletMain->GetAddressBookEntry(dest);
     // Check if we have a new address or an updated label
-    if (mi == wallet->mapAddressBook.end()) {
+    if (!optAdd) {
         return wallet->SetAddressBook(dest, strName, strPurpose);
-    } else if (mi->second.name != strName) {
+    } else if (optAdd->name != strName) {
         return wallet->SetAddressBook(dest, strName, ""); // "" means don't change purpose
```
Comment (random-zebra) [inline src/qt/walletmodel.cpp:299]: This makes the LOCK at line 292 not needed now. Both `GetABEntry` and `SetAddressBook` lock cs_wallet when needed.

File: src/wallet/rpcwallet.cpp [hunk 6]
```diff
@@ -591,9 +593,9 @@ UniValue delegatorremove(const JSONRPCRequest& request)
...
     {
         LOCK(pwalletMain->cs_wallet);
-        std::map<CTxDestination, AddressBook::CAddressBookData>::iterator mi = pwalletMain->mapAddressBook.find(dest);
-        if (mi != pwalletMain->mapAddressBook.end()) {
-            label = mi->second.name;
+        auto optAdd = pwalletMain->GetAddressBookEntry(dest);
+        if (optAdd) {
+            label = optAdd->name;
         }
     }
```
Comment (random-zebra) [inline src/wallet/rpcwallet.cpp:600]: no need for the LOCK now

**Review by random-zebra — COMMENTED** [review]
Sorry for multiple reviews... but have found something else, about the recursive locking being introduced here. See inline comments (there could be other instances too).

**Comment by furszy** [comment]
updated solving all of the comments 👍

**Comment by furszy** [comment]
rebased on master due conflicts with recently merged #1853 plus cleaned a previous refactor that wasn't needed anymore in `ListaddressesForPurpose` pointed by zebra.

Ready for review.

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -2087,7 +2086,7 @@ UniValue ListReceived(const UniValue& params, bool by_label)
         }
 
         if (by_label) {
-            tallyitem& _item = label_tally[label];
+            tallyitem &_item = label_tally[label];
```
Comment (random-zebra) [inline src/wallet/rpcwallet.cpp:2089]: (e0ffc983ce0af18cdd1584176a7b7b5b581d8d5d)
these were fine. Fix your IDE auto-correction tool :stuck_out_tongue:

File: src/wallet/wallet.h [hunk 8]
```diff
@@ -250,11 +250,13 @@ class CAddressBookIterator
     explicit CAddressBookIterator(std::map<CTxDestination, AddressBook::CAddressBookData>& _map) : map(_map), it(_map.begin()), itEnd(_map.end()) {}
     CTxDestination GetKey() { return it->first; }
     AddressBook::CAddressBookData GetValue() { return it->second; }
-    bool HasNext() { return it != itEnd; }
+
+    bool IsValid() { return it != itEnd && it != map.end(); }
```
Comment (random-zebra) [inline src/wallet/wallet.h:254]: (e0ffc983ce0af18cdd1584176a7b7b5b581d8d5d)
This change confuses me.
If we set the filter and advance with `Next()`, how can we get past `itEnd` and reach `map.end()` (if they are different)?

**Review by random-zebra — DISMISSED** [review]
Looks good.
There is still some recursive locking (e.g. `HasAddressBook` or `GetNameForAddressBookEntry`), but those can be fixed later.
Same as a couple comments left inline.
ACK e0ffc983ce0af18cdd1584176a7b7b5b581d8d5d

File: src/wallet/wallet.h (hunk 8 above)
Comment (furszy) [reply src/wallet/wallet.h:254]: With the current code, we can't yeah. I added it previously with another subfiltering logic in mind (when the `SetFilter` wasn't a `map.find` only). can remove it 👍 .

**Comment by furszy** [comment]
Done, updated the two lines changes. A quick git diff and can be re acked.

**Review by random-zebra — APPROVED** [review]
re-utACK f7e997fa88cad4cd3502b4b372667928f62a38bb

**Review by Fuzzbawls — APPROVED** [review]
utACK f7e997fa88cad4cd3502b4b372667928f62a38bb

**Pull request merged**